### DISCUSSION TITLE
Let's talk DoEvents a sec...

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#1119] [Ideas]
I haven't seen this discussed here as yet, and thought now may be a great opportunity to bring it up....along with a question...or two.

1) When is DoEvents slated to be implemented?
2) Can we - perhaps - enhance DoEvents to be...more choosy/precise in what it allows? (perhaps we DON'T want to allow just *everything* to proceed?)

The scenario in which this second question arises is Office (read: Excel) Automation. While automating an Excel application instance, there are times when it is beneficial to "DoEvents" and let Excel settle its outstanding stuff before proceeding (like especially when dealing with a recalcitrant 1004 Automation error from Excel, for one example). However, there may be events _outside of Excel_ that would be better to NOT have proceed (like, say someone from the OS/Explorer double-clicking on another .xls file and firing up another Excel window)

So, I was imagining an event filter (optional, of course) parameter to DoEvents, so:

`
DoEvents (Optional FilterEventCriteria() as variant)  'or, think: ParamArray here
`
...perhaps?
Where the FilterEventCtireria may be something like a (list of) Process IDs, or an Application handle, or a window handle....?
(I'm just thinking aloud here on this, folks.)

The point being to allow only _some_ things to happen - or, alternately it may be desirable to allow everything else EXCEPT this (list of whatever) to proceed...?
I could see the need for both options in certain conditions.

Right now, all we have is the shotgun-approach to DoEvents....and I'm wondering if we couldn't get something more....nuanced, also.

Now, I get that there are myriad issues with this in terms of "how to we restrict what windows does in its message loop handlers?"
to which I have no immediate answers, but thought that placing the question before all of you may yield some useful thoughts.

**Comment by mburns08109** [comment]
Here is one thought for that last question:
is there a reason we couldn't "roll our own" limited-purpose Windows message pump for the purposes of this function?

**Comment by bclothier** [comment]
TBH, I'm inclined to just slap it with a warning that it's not recommended, especially in the context of tB. The whole reason for `DoEvents` is due to the underlying architectural issues with VBx (e.g. running on a single thread, blocking messages from being pumped, etc.) and re-think the approach altogether. That is especially true in tB where 1) multithreading is already possible and 2) it's a compiled application that doesn't have to share its main thread with others unlike VBA project which must have the main thread or else. 

Therefore, there will be better solutions available to us that wouldn't be feasible under VBx. Enhancing `DoEvents `would only end up as a dead end alley. So I see no point in adding lipstick to a pig here. Just shoot it dead and take its belly, smoke it and enjoy ourselves some bacon.

**Comment by dmrvb** [comment]
I use DoEvents and timers extensively to make single threaded VB6 apps feel a bit multithreaded. In particular I need software to run in real time but sometimes do a big calculation whilst still responding to **some** user clicks. A better DoEvents would be great, but an easy to use multi-threading would be even better. And the VB6 (blocking) menu system is a nightmare.....what is tB going to do about the menu system??

**Comment by Kr00l** [comment]
There are a lot of "alternative" DoEvents in the wild for VB6 which just do clear the message pump via PeekMessage.
So, IMO tB does not need to solve everything as super fine grained out of the box.
That's where a TwinPackage might be handy to provide more sophisticated DoEvents functions..

**Comment by rexxitall** [comment]
Do events is the most stupid thing ever ! Its there cause they was too stupid that time to make a real good multitasking task scheduler. Which is BTW not that hard to implement. For compatibilty DoEvents ok.  But poll events - NO ! I really love multitasking program(ing)s since OS/2. Somehow there was never such a thing. The OS calls your routine as soon as possible. That was it. Same should be inside TB. At least one thread / task for GUI and friends and another for calculating things. That works quite well. Even if a civil-engineer program it :) Just in case someone wants to make it better then Microsoft. I got it mad to put in every better for next slope a doEvents to remind the genius behind microsoft to get awake. The disadvantage on this is. If the things goes wrong. The program gets frozen. Cause no one returns the control. Even in the good old days in REXX and OS/2 i was at least be able to watch about the thread which i started. And if i feel that there is something worn. I could stop the BS :) In REXX !  :) I really love the TB philosophy. But please do not  program blue screens just for that :)

**Comment by VBFNiya** [comment]
Quite frankly DoEvents is evil and everything should be done to discourage it's use. I know there are plans to make multi-threading a first class citizen in tB and when that happens, DoEvents would be completely unnecessary. There is **no** problem and I mean none where DoEvents is a better solution than multi-threading. DoEvents in my opinion is just straight garbage because it encourages very poor design decisions that lead to buggy code.

**Reply by mburns08109** [reply]
You are just plain wrong on this because you are forgetting the context of the question.
The Context was within the scope of an ActiveX Automation scenario, where DoEvents does indeed have operational significance irrespective of whether the tB task itself was using a different non-apartment threading model or not.

It doesn't matter how much you might Hate DoEvents or apartment threading models here. They are an established fact for ActiveX automation scenarios, and no amount of grousing nor grinding of teeth over it will make that go away.

So can we get back to the point of my question, please?  (sorry @VBFNiya please don't take that as a personal attack sort of reply - I meant it as a more generic response, but errantly posted it as a reply to you.)

**Reply by wqweto** [reply]
There are no “ActiveX Automation scenarios with DoEvent you are talking about. No one outside VBx worlds knows about DoEvents anyway but what usually is discussed is message pumps and spinning messages in STA. Why would any architect of the underlying COM burden anything there with some fringe VBx function which can be reimplemented in 5 lines of code? DoEvents is not some magic but poorly named SpinMessages routine.

Using DoEvents for “multi-threading” is in itself a terrible idea. What is this? Some kind of cooperarive multi-tasking? Yeilding to other “tasks” with DoEvent? Doing this manually would be a re-entrancy hell and have never seen this implemented methodically.

Only viable pure VBx reason for calling DoEvent would be UI refresh and cancellation button impl on a long running modal progress indication dialog.

**Reply by VBFNiya** [reply]
> Using DoEvents for “multi-threading” is in itself a terrible idea. What is this? Some kind of cooperarive multi-tasking? Yeilding to other “tasks” with DoEvent? Doing this manually would be a re-entrancy hell and have never seen this implemented methodically.

Oh I know this pain. It was a deciding factor in my decision to leave VB6 in favor of VB.Net. I wrote a Winsock application in VB6 and gave it the **appearance** of performing asynchronous downloads by using DoEvents to keep the UI responsive. It was an absolute nightmare to debug. There were endless reentrancy problems. The experience was so traumatic that it finally gave me the courage to leave VB6. I never wanted to experience that kind of pain again.

**Reply by bclothier** [reply]
I previously mentioned that a `DoEvents` ought to merit a compiler warning. However in tB, we have both STA and MTA projects. VBx was always STA* so a tB MTA project is an entirely new thing. I would probably want to make usage of `DoEvents` an error rather than a warning in that context. 

---

<sup>* There was a half-hearted attempt at MT'd VB6 projects provided by Microsoft but I think that was never seriously taken up? Not sure I've seen any serious implementation of MT'd  VB6 in the wild using those project templates.</sup>
